src/Workspaces/Core/Portable/Differencing/LongestCommonSubsequence.cs

```diff
@@ -11,73 +11,150 @@ namespace Microsoft.CodeAnalysis.Differencing
                 {
-                    j--;
+                    return array[index + offset];
                 }
-                else
+                set
                 {
-                    i--;
-                    j--;
-                    yield return new KeyValuePair<int, int>(i, j);
+                    array[index + offset] = value;
                 }
             }
         }
 
-        protected IEnumerable<SequenceEdit> GetEdits(TSequence oldSequence, int oldLength, TSequence newSequence, int newLength)
+        protected abstract bool ItemsEqual(TSequence oldSequence, int oldIndex, TSequence newSequence, int newIndex);
+
+        protected IEnumerable<KeyValuePair<int, int>> GetMatchingPairs(TSequence oldSequence, int oldLength, TSequence newSequence, int newLength)
         {
-            int[,] d = ComputeCostMatrix(oldSequence, oldLength, newSequence, newLength);
-            int i = oldLength;
-            int j = newLength;
+            Stack<VArray> stackOfVs = ComputeEditPaths(oldSequence, oldLength, newSequence, newLength);
+
+            int x = oldLength;
+            int y = newLength;
 
-            while (i != 0 && j != 0)
+            for (int d = stackOfVs.Count - 1; x > 0 || y > 0; d--)
             {
-                if (d[i, j] == d[i - 1, j] + DeleteCost)
+                VArray currentV = stackOfVs.Pop();
+                int k = x - y;
+
+                // "snake" == single delete or insert followed by 0 or more diagonals
+                // snake end point is in V
+                int yEnd = currentV[k];
+                int xEnd = yEnd + k;
+
+                // does the snake first go down (insert) or right(delete)?
+                bool right = (k == d || (k != -d && currentV[k - 1] > currentV[k + 1]));
+                int kPrev = right ? k - 1 : k + 1;
+
+                // snake start point
+                int yStart = currentV[kPrev];
+                int xStart = yStart + kPrev;
+
+                // snake mid point
+                int yMid = right ? yStart : yStart + 1;
+                int xMid = yMid + k;
+
+                // return the matching pairs between (xMid, yMid) and (xEnd, yEnd) = diagonal part of the snake
+                while (xEnd > xMid)
                 {
-                    i--;
-                    yield return new SequenceEdit(i, -1);
+                    Debug.Assert(yEnd > yMid);
+                    xEnd--;
+                    yEnd--;
+                    yield return new KeyValuePair<int, int>(xEnd, yEnd);
                 }
-                else if (d[i, j] == d[i, j - 1] + InsertCost)
+
+                x = xStart;
+                y = yStart;
+            }
+
+        }
+
+        protected IEnumerable<SequenceEdit> GetEdits(TSequence oldSequence, int oldLength, TSequence newSequence, int newLength)
+        {
+            Stack<VArray> stackOfVs = ComputeEditPaths(oldSequence, oldLength, newSequence, newLength);
+
+            int x = oldLength;
```

Comment:
I wonder if we can somehow avoid duplicating the code between ```GetEdits``` and ```GetMatchingParts```. It seems that  a simple generalization to ```GetEdits``` could make it so that ```GetMatchingParts``` is a specialization of it. Adding a ```bool``` parameter that skips over lines 142-153. Would that work?